runtime/common/MeasureCounts.cpp

```diff
@@ -345,7 +345,31 @@ bool sample_result::has_expectation(const std::string_view registerName) {
   return iter->second.expectationValue.has_value();
 }
 
-double sample_result::exp_val_z(const std::string_view registerName) {
+double sample_result::expectation(const std::string_view registerName) {
+  double aver = 0.0;
+  auto iter = sampleResults.find(registerName.data());
+  if (iter == sampleResults.end())
+    return 0.0;
+
+  if (iter->second.expectationValue.has_value())
+    return iter->second.expectationValue.value();
+
+  auto counts = iter->second.counts;
+  for (auto &kv : counts) {
+    auto par = has_even_parity(kv.first);
+    auto p = probability(kv.first, registerName);
+    if (!par) {
+      p = -p;
+    }
+    aver += p;
+  }
+
+  return aver;
+}
+
+[[deprecated("`exp_val_z()` is deprecated. Use `expectation()` with the same "
+             "argument structure.")]] double
+sample_result::exp_val_z(const std::string_view registerName) {
```

Comment:
I could be wrong about this, but doesn't the `[[deprecated]]` tag have to be in the header file to produce the warning if users try to use it?

Reply:
I don't believe so -- I can produce the error as follows with this branch

```
import cudaq

kernel = cudaq.make_kernel()
qubits = kernel.qalloc(3)
kernel.x(qubits)

result = cudaq.sample(kernel)
expectation_good = result.expectation()
expectation_bad = result.expectation_z()
```

prints
```
/workspaces/cuda-quantum/build/test.py:9: DeprecationWarning: expectation_z() is deprecated, use expectation() with the same argument structure.
  expectation_bad = result.expectation_z()
```

Reply:
Is the same true for C++?

Reply:
For example, if I checkout this PR and revert `test/AST-Quake/simple.cpp` back to the original `exp_val_z()` and run `nvq++ test/AST-Quake/simple.cpp`, I get no deprecation warning. I also get no deprecation warning at runtime either.

Reply:
You're absolutely correct, thank you for double checking me on that. My brain defaulted to python mode there for a second -- fixing it now 🤦

Reply:
Fixed with the most recent commit. Just to clarify though, we only expect a compile-time error here, right? There's no runtime warning as-is.
```
$ nvq++ ../test/AST-Quake/simple.cpp -o out.x
../test/AST-Quake/simple.cpp:95:31: warning: 'exp_val_z' is deprecated: `exp_val_z()` is deprecated. Use `expectation()` with the same argument structure. [-Wdeprecated-declarations]
  printf("Exp: %lf\n", counts.exp_val_z());
                              ^
/usr/local/cudaq/include/common/MeasureCounts.h:182:5: note: 'exp_val_z' has been explicitly marked deprecated here
  [[deprecated("`exp_val_z()` is deprecated. Use `expectation()` with the same "
    ^
1 warning generated.
```

Reply:
> Just to clarify though, we only expect a compile-time error here, right?

Yes, I agree. Thanks!